src/order_cmd.cpp

```diff
@@ -312,39 +293,24 @@ void OrderList::RecalculateTimetableDuration()
  */
 void OrderList::FreeChain(bool keep_orderlist)
 {
-	Order *next;
-	for (Order *o = this->first; o != nullptr; o = next) {
-		next = o->next;
-		delete o;
+	/* We can visit oil rigs and buoys that are not our own. They will be shown in
+	 * the list of stations. So, we need to invalidate that window if needed. */
+	for (Order &order: this->orders) {
+		if (order.IsType(OT_GOTO_STATION) || order.IsType(OT_GOTO_WAYPOINT)) {
+			BaseStation *bs = BaseStation::GetIfValid(order.GetDestination().ToStationID());
+			if (bs != nullptr && bs->owner == OWNER_NONE) InvalidateWindowClassesData(WC_STATION_LIST, 0);
```

Comment:
We don't need to invalidate the station list multiple times, right?
```suggestion
			if (bs != nullptr && bs->owner == OWNER_NONE) {
				InvalidateWindowClassesData(WC_STATION_LIST, 0);
				break;
			}
```

Reply:
True, seems like a change that should be done outside this PR.